src/gravel/controllers/nodes/deployment.py

```diff
@@ -124,3 +148,120 @@ def mark_ready(self) -> None:
+
+    @property
+    def state(self) -> DeploymentState:
+        return self._state
+
+    async def join(
+        self,
+        leader_address: str,
+        token: str,
+        uuid: UUID,
+        hostname: str,
+        address: str
+    ) -> bool:
+        logger.debug(
+            f"join > with leader {leader_address}, token: {token}"
+        )
+
+        assert self._state
+        assert hostname
+        assert address
+
+        if self._state.bootstrapping:
+            raise NodeBootstrappingError()
+        elif self._state.deployed:
+            raise NodeHasBeenDeployedError()
+        elif self._state.joining:
+            raise NodeAlreadyJoiningError()
+        elif self._state.ready:
+            raise NodeHasJoinedError()
+        assert self._state.nostage
+
+        uri: str = f"ws://{leader_address}/api/nodes/ws"
+        conn = await self._connmgr.connect(uri)
+        logger.debug(f"join > conn: {conn}")
+
+        joinmsg = JoinMessageModel(
+            uuid=uuid,
+            hostname=hostname,
+            address=address,
+            token=token
+        )
+        msg = MessageModel(type=MessageTypeEnum.JOIN, data=joinmsg.dict())
+        await conn.send(msg)
+
+        self._state.mark_join()
+
+        reply: MessageModel = await conn.receive()
+        logger.debug(f"join > recv: {reply}")
+        if reply.type == MessageTypeEnum.ERROR:
+            errmsg = ErrorMessageModel.parse_obj(reply.data)
+            logger.error(f"join > error: {errmsg.what}")
+            await conn.close()
+            self._state.mark_error()
+            return False
+
+        assert reply.type == MessageTypeEnum.WELCOME
+        welcome = WelcomeMessageModel.parse_obj(reply.data)
+        assert welcome.pubkey
+        assert welcome.cephconf
+        assert welcome.keyring
+        assert welcome.etcd_peer
+
+        my_url: str = \
+            f"{hostname}=http://{address}:2380"
+        initial_cluster: str = f"{welcome.etcd_peer},{my_url}"
+        await spawn_etcd(
+            self._gstate,
+            new=False,
+            token=None,
+            hostname=hostname,
+            address=address,
+            initial_cluster=initial_cluster
+        )
+
+        authorized_keys: Path = Path("/root/.ssh/authorized_keys")
+        if not authorized_keys.parent.exists():
+            authorized_keys.parent.mkdir(0o700)
+        with authorized_keys.open("a") as fd:
+            fd.writelines([welcome.pubkey])
+            logger.debug(f"join > wrote pubkey to {authorized_keys}")
+
+        cephconf_path: Path = Path("/etc/ceph/ceph.conf")
+        keyring_path: Path = Path("/etc/ceph/ceph.client.admin.keyring")
+        if not cephconf_path.parent.exists():
+            cephconf_path.parent.mkdir(0o755)
+        cephconf_path.write_text(welcome.cephconf)
+        keyring_path.write_text(welcome.keyring)
+        keyring_path.chmod(0o600)
+        cephconf_path.chmod(0o644)
```

Comment:
(nit) wondering if this belongs in a Config class (or similar)?

Reply:
it probably belongs somewhere else, but I don't think it would be in the Config class. Then again, I don't expect this to be used anywhere else, so I'm not particularly bothered by having it here :shrug: